fs/nvs: hardening against bad offset in close_ate

nvlsianpu opened the pull request:
If closing ate is valid (crc is ok) and close_ate.offset
has not allowed value GC can't be performed.
Such sector might be treated as empty as it is remains after
disturbed nvs write.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
fs/nvs: hardening against bad offset in close_ate

If closing ate is valid (crc is ok) and close_ate.offset
has not allowed value GC can't be performed straight forward.
Such sector might be treated as to by recovery first as it might
be remains after a disturbed nvs write.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -484,8 +484,17 @@ static int nvs_gc(struct nvs_fs *fs)
+			 * doesn't contain any data.
+			 * Consider this sector as empty.
+			 */
+			goto clean_up;
```
I might consider nvs_recovery_last_ate() instead.

nvlsianpu committed:
fs/nvs: allow recovery sectors with big ate gaps

Encountered issue when sector contained big gaps in metadata.
So far NVS has been allowed to recovery ate's until it encounter
an empty ate.

This patch makes gaps optionally allowed as introduces seeking over
entire sector.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

Laczen commented:
@nvlsianpu, I don't know what bug you have discovered. I don't think it is a good idea to go trough the complete sector to find the last valid ate. The data that is before the ate's can contain a something that is equivalent to a valid ate, but it is not, it is data. Therefore the end of the ate should be an empty entry or a ate pointed to by the sector closing ate. There should be no empty ate's  in the ate table, even if the write goes wrong. So an empty ate that is found is the end of the ate's.

nvlsianpu committed:
fs/nvs: Add re-write and write check options

Added option for check whether what was written match to what
was requested to be write.

Added option for try rewrite in case flash content doesn't match.
This option can be enabled only on memories which allow writes to
same location at last twice.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
fs/nvs: serve empty-closed sector

For a sector it is possible that close_ate.offset == stop_addr as
the results of closing the sector while it contained only corrupted
written data.
It is possible corner case, and such sector should be treated as empty
by GC.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
It was encountered issue when NVS iterates over the end of the device flash as the result of too big offset in the sector closing ATE. it is possible that `close_ate.offset == stop_addr` as the results of closing the sector while it contained only corrupted written data. GC procedure didn't serve such corner case.

I get to conclusion that sector was recognized as containing lot of garbage's in results of a few ineffective NVS writes in a row before the successful write. Than after reset, NVS had recognized that this sector was occupied by garbage's almost up to the end (and no data). So on the next write it was ease to trigger sector rotation.

mekugi commented on subsys/fs/nvs/Kconfig:
```diff
@@ -14,4 +14,26 @@ module = NVS
+	  This option can be enabled only on memories which allow writes to
+	  same location at las twice.
+
+config CONFIG_NVS_ALLOW_ATE_GAPS
```
```suggestion
config NVS_ALLOW_ATE_GAPS
```

Laczen commented:
@nvlsianu,
a. The write/rewrite option does not belong in nvs. If required it should be in the flash driver. 
b. ATE gaps should not be allowed, they should never be there.

Laczen commented on subsys/fs/nvs/Kconfig:
```diff
@@ -14,4 +14,26 @@ module = NVS
+	  writes on devices on which same flash location rewrite attempts are
+	  forbidden or flash malfunction might cause ATE being empty.
+
 endif # NVS
```
None of these options should be added

nvlsianpu commented:
>a. The write/rewrite option does not belong in nvs. If required it should be in the flash driver.
some devices allow rewrites, some not, and know one device on which it causes fault. this made possible solution some more complex than I expected. Will prepare other solution then.

>b. ATE gaps should not be allowed, they should never be there.

In general I had similar opinion, but I considered following case:
On some flash device multiple writes to same minimal-write-block are forbidden. If error will result in having ATE field empty, we can't even made this ATE invalidated. What I already have done allow having a recovery mechanism (post factum). And I made this optional.

I can consider another path which will cause nvs sector rotation in such a case this issue just happen.

Laczen commented:
> @Laczen
> 
> > a. The write/rewrite option does not belong in nvs. If required it should be in the flash driver.
> 
> some devices allow rewrites, some not, and know one device on which it causes fault. this made possible solution some more complex than I expected. Will prepare other solution then.
> 
nvs never uses rewrites. The device that causes fault should be studied. What could be the problem is that the end of life has been reached and that flash write fails, this would be a problem for all subsystems that use flash and therefore should be handled in the flash driver.

> > b. ATE gaps should not be allowed, they should never be there.
> 
> In general I had similar opinion, but I considered following case:
> On some flash device multiple writes to same minimal-write-block are forbidden. If error will result in having ATE field empty, we can't even made this ATE invalidated. What I already have done allow having a recovery mechanism (post factum). And I made this optional.
> 
> I can consider another path which will cause nvs sector rotation in such a case this issue just happen.
As nvs never uses rewrites this could only be caused by a trial write that was performed and not completed. In that case a empty ate could no longer be written. If this is true than those devices at the moment would not be supported by nvs. It would be required to introduce a new kind of detection for the end of ate's: this could be empty for existing or "writeable" for these special devices.

de-nordic commented on subsys/fs/nvs/Kconfig:
```diff
@@ -14,4 +14,26 @@ module = NVS
+	help
+	  Retry write operation if flash content doesn't match to data written.
+	  This option can be enabled only on memories which allow writes to
+	  same location at las twice.
```
at las->at least?

de-nordic commented on subsys/fs/nvs/Kconfig:
```diff
@@ -14,4 +14,26 @@ module = NVS
+	  same location at las twice.
+
+config NVS_ALLOW_ATE_GAPS
+	bool "Allow big gaps in ATE area"
```
Maybe change the title to "Scan presumed empty ATEs to recover last ate" or something like this, so that the title would more closely describe the purpose.

de-nordic commented on subsys/fs/nvs/Kconfig:
```diff
@@ -14,4 +14,26 @@ module = NVS
+
+config NVS_WRITE_RETRY
+	bool "Retry write"
+	depends on NVS_WRITE_CHECK
```
Shouldn't that be rather `select`?

de-nordic commented on subsys/fs/nvs/nvs.c:
```diff
@@ -73,6 +77,35 @@ static int nvs_flash_al_wrt(struct nvs_fs *fs, uint32_t addr, const void *data,
+{
+	int rc;
+
+#ifdef CONFIG_NVS_WRITE_RETRY
+	for (int i = 0; i < 2; i++) {
+#endif
```
Maybe do `#else` with just opening bracket. This way the ifdef in 103-105 could be avoided, although I do not know whether style allows that.

de-nordic commented:
> 
> nvs never uses rewrites. The device that causes fault should be studied. What could be the problem is that the end of life has been reached and that flash write fails, this would be a problem for all subsystems that use flash and therefore should be handled in the flash driver.

As far as I can tell from the code the rewrite happens after flash write function fails and  the function itself does not verify write, it is API call directed to drive; the causes of failure may be other than actually failing to flip bits. So repeating the write in such case should be ok although some warning should be reported. 
On the other side If we allow re-writes the re-read should be mandatory, not a Kconfig option,  to verify data and when re-read fails the comparison, the sector should be closed with special mark for gc/clear treatment or as bad-sector and write should be attempted to other sector.

de-nordic commented on subsys/fs/nvs/nvs.c:
```diff
@@ -73,6 +77,35 @@ static int nvs_flash_al_wrt(struct nvs_fs *fs, uint32_t addr, const void *data,
+#ifdef CONFIG_NVS_WRITE_RETRY
+	for (int i = 0; i < 2; i++) {
+#endif
+		rc = _nvs_flash_al_wrt(fs, addr, data, len);
```
There should be some logging if this fails, in case when re-writes are enabled.

de-nordic commented on subsys/fs/nvs/nvs.c:
```diff
@@ -73,6 +77,35 @@ static int nvs_flash_al_wrt(struct nvs_fs *fs, uint32_t addr, const void *data,
+		rc = _nvs_flash_al_wrt(fs, addr, data, len);
+#ifdef CONFIG_NVS_WRITE_CHECK
+		if (rc == 0) {
+			rc = nvs_flash_block_cmp(fs, addr, data, len);
```
Same here, some logging.
